In [58]:

#Nama Ricki Saputra
#I.2310394
import kagglehub

# Download latest version
path = kagglehub.dataset_download("erqizhou/students-data-analysis")

print("Path to dataset files:", path)

Using Colab cache for faster access to the 'students-data-analysis' dataset.
Path to dataset files: /kaggle/input/students-data-analysis


In [60]:
import os

dataset_path = '/kaggle/input/students-data-analysis'
files = os.listdir(dataset_path)

print("Daftar File dalam Folder:")
for f in files:
    print("-", f)

Daftar File dalam Folder:
- Students data.csv


In [ ]:
import pandas as pd
data_files = os.path.join(dataset_path, 'Students data.csv')
data = pd.read_csv(data_files)
data

ID class  gender  race    GPA  Algebra  Calculus1  Calculus2  \
0    1141     A    male     1  73.47       64         81         87   
1    1142     A  female     1  71.22       57         50         51   
2    1143     A  female     2  74.56       47         48         71   
3    1144     A  female     1  72.89       46         72         38   
4    1145     A  female     1  70.11       49         45         63   
..    ...   ...     ...   ...    ...      ...        ...        ...   
100  1241     A  female     1  88.34       87         83         92   
101  1242     B    male     1  89.84       98         77         95   
102  1243     B    male     1  88.82       83         80         91   
103  1244     A    male     1  86.60       92         82         91   
104  1245     A    male     1  93.71       93         97         99   

     Statistics  Probability  Measure  Functional_analysis from1 from2 from3  \
0            60           74       71                   60     A     A     A   
1            51           55       62                   61     B     A     A   
2            60           61       68                   64     C     A     A   
3            60           29       54                   51     D     A     A   
4            60           66       66                   61     E     A     A   
..          ...          ...      ...                  ...   ...   ...   ...   
100          98           93       86                   90     M     B     A   
101          98           96       88                  100     A     B     A   
102          98           93       95                   71     T     B     A   
103          99           94       82                   78     S     B     A   
104         100           97       90                   90     K     B     A   

     from4  y  
0        3  0  
1        2  0  
2        0  1  
3        0  0  
4        0  0  
..     ... ..  
100      0  1  
101      0  1  
102      0  2  
103      0  2  
104      0  2  

[105 rows x 17 columns]

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 105 entries, 0 to 104
Data columns (total 17 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   ID                   105 non-null    int64  
 1   class                105 non-null    object 
 2   gender               105 non-null    object 
 3   race                 105 non-null    int64  
 4   GPA                  105 non-null    float64
 5   Algebra              105 non-null    int64  
 6   Calculus1            105 non-null    int64  
 7   Calculus2            105 non-null    int64  
 8   Statistics           105 non-null    int64  
 9   Probability          105 non-null    int64  
 10  Measure              105 non-null    int64  
 11  Functional_analysis  105 non-null    int64  
 12  from1                105 non-null    object 
 13  from2                105 non-null    object 
 14  from3                105 non-null    object 
 15  from4                105 non-null    int

##Random Forest untuk 5 fitur

In [62]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import mean_squared_error, r2_score, accuracy_score

In [69]:
x = data[['Algebra', 'Calculus1','Calculus2','Statistics','Probability']]
y = data[['y']]

In [66]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2)

In [70]:
model = RandomForestClassifier()
model.fit(x_train, y_train)

y_pred_train = model.predict(x_train)
y_pred_test = model.predict(x_test)

akurasi_training = accuracy_score(y_train, y_pred_train)
akurasi_testing = accuracy_score(y_test, y_pred_test)

print(f"training = {akurasi_training:.2%}")
print(f"testing = {akurasi_testing:.2%}")

/usr/local/lib/python3.12/dist-packages/sklearn/base.py:1389: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return fit_method(estimator, *args, **kwargs)


training = 100.00%
testing = 57.14%


###Dari Hasil training i atas Akurasi menunjukkan 100% dengan akurasi testing sebesar 57% sebelum di lakukan reduksi dengan PCA

In [71]:
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2)

x = data[['Algebra', 'Calculus1','Calculus2','Statistics','Probability','Measure']]
y = data[['y']]

In [72]:
scaler = StandardScaler()
x_scaled = scaler.fit_transform(x)

pca = PCA(n_components=5)
x_pca = pca.fit_transform(x_scaled)

In [73]:

model = RandomForestClassifier()
model.fit(x_train, y_train)

y_pred_train = model.predict(x_train)
y_pred_test = model.predict(x_test)

akurasi_training = accuracy_score(y_train, y_pred_train)
akurasi_testing = accuracy_score(y_test, y_pred_test)

print(f"Hasil akurasi training = {akurasi_training:.2%}")
print(f"Hasil akurasi testing = {akurasi_testing:.2%}")

/usr/local/lib/python3.12/dist-packages/sklearn/base.py:1389: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return fit_method(estimator, *args, **kwargs)


Hasil akurasi training = 100.00%
Hasil akurasi testing = 47.62%


###Dari hasil training di atas,
hasil training tetap sama di 100% namun akurasi testing turun menjadi 47% setelah dilakukan redukasi Menggunakan 5 fitur